scipy/special/_cosine.c

```diff
@@ -0,0 +1,292 @@
+/*
+ *  The functions
+ *
+ *    cosine_cdf
+ *    cosine_invcdf
+ *
+ *  defined here are the kernels for the ufuncs
+ *
+ *    _cosine_cdf
+ *    _cosine_invcdf
+ *
+ *  defined in scipy.special._ufuncs.
+ *
+ *  The ufuncs are used by the class scipy.stats.cosine_gen.
+ */
+
+#include <stdio.h>
+#include <math.h>
+
+#include "cephes/polevl.h"
+
+
+//
+// p and q (below) are the coefficients in the numerator and denominator
+// polynomials (resp.) of the Pade approximation of
+//     f(x) = (pi + x + sin(x))/(2*pi)
+// at x=-pi.  The coefficients are ordered from lowest degree to highest.
+// These values are used in the function cosine_cdf_pade_approx_at_neg_pi(x).
+//
+// These coefficients can be derived by using mpmath as follows:
+//
+//    import mpmath
+//
+//    def f(x):
+//        return (mpmath.pi + x + mpmath.sin(x)) / (2*mpmath.pi)
+//
+//    # Note: 40 digits might be overkill; a few more digits than the default
+//    # might be sufficient.
+//    mpmath.mp.dps = 40
+//    ts = mpmath.taylor(f, -mpmath.pi, 20)
+//    p, q = mpmath.pade(ts, 9, 10)
+//
+// (A python script with that code is in special/_precompute/cosine_cdf.py.)
+//
+// The following are the values after converting to 64 bit floating point:
+// p = [0.0,
+//      0.0,
+//      0.0,
+//      0.026525823848649224,
+//      0.0,
+//      -0.0007883197097740538,
+//      0.0,
+//      1.0235408442872927e-05,
+//      0.0,
+//      -3.8360369451359084e-08]
+// q = [1.0,
+//      0.0,
+//      0.020281047093125535,
+//      0.0,
+//      0.00020944197182753272,
+//      0.0,
+//      1.4162345851873058e-06,
+//      0.0,
+//      6.498171564823105e-09,
+//      0.0,
+//      1.6955280904096042e-11]
+//
+
+
+//
+// Compute the CDF of the standard cosine distribution for x close to but
+// not less than -π.  A Pade approximant is used to avoid the loss of
+// precision that occurs in the formula 1/2 + (x + sin(x))/(2*pi) when
+// x is near -π.
+//
+static
+double cosine_cdf_pade_approx_at_neg_pi(double x)
+{
+    double h, h2, h3;
+    double numer, denom;
+    double numer_coeffs[] = {-3.8360369451359084e-08,
+                             1.0235408442872927e-05,
+                             -0.0007883197097740538,
+                             0.026525823848649224};
+    double denom_coeffs[] = {1.6955280904096042e-11,
+                             6.498171564823105e-09,
+                             1.4162345851873058e-06,
+                             0.00020944197182753272,
+                             0.020281047093125535,
+                             1.0};
+
+    // M_PI is not exactly π.  In fact, float64(π - M_PI) is
+    // 1.2246467991473532e-16.  h is supposed to be x + π, so to compute
+    // h accurately, we write the calculation as:
+    h = (x + M_PI) + 1.2246467991473532e-16;
```

Comment:
Just to check myself: the idea is that even though (M_PI + 1.2246467991473532e-16) may not be distinguishable from 1.2246467991473532e-16 (at least not with much precision), this stuff is specifically written for when (x + M_PI) is small enough that the addition of 1.2246467991473532e-16 can be resolved and makes a difference.

Reply:
Yes, exactly.